### PR TITLE
[LTX] Add bf8 support

### PR DESCRIPTION
**Pull request by smartonTT**
## What

main's LTX pipeline was **bf16-only** — it has no DiT-linear quant path. This adds an **optional** bf8
quant tier: a new `quant_config.py`, the activation-cast plumbing in `linear.py`, and the bf8-consuming
refactor in `attention_ltx.py`/`transformer_ltx.py` (gather-once + activation-dtype), wired into
`pipeline_ltx.py`. **The default stays bf16** — `LTX_QUANT_DEFAULT=""` applies no quant config, so the
default path is byte-for-byte the bf16 baseline. Opt into the faster tier with `LTX_QUANT=all_bf8_lofi`. `LTX_QUANT_ACTIVATIONS` (default on)
narrows every bf8 activation that crosses the fabric — the linear casts and the ring-SDPA Q/K/V
inputs — under one flag, so a run can't narrow the linears while leaving SDPA at bf16 (the SDPA
input cast alone is ~2% e2e, measured). No shared base classes touched.

**bf16-activations option (under evaluation):** set `LTX_QUANT_ACTIVATIONS=0` to keep activations — the
linear casts, ring-SDPA inputs, and output pins — at **bf16** while weights stay bf8. It's ~7% slower
(6.8 s vs 6.35 s, SEED=10 warm — bf16 collectives move 2× the bytes), but appears better-behaved: the
render tracks the bf16 model more closely (semantics/scene). Kept as an option pending fuller quality
evaluation; bf8 activations remain the default.

**bf4 evaluated, not included:** 4-bit **weights** give no speedup — this workload isn't
weight-bandwidth-bound, so bf4-weights render at the same ~6.4 s as bf8 (≈ bf8 quality). 4-bit
**activations** also give no speedup (~6.3 s) and **badly degrade quality** — hazy, washed-out,
low-contrast (MUSIQ −21, NIMA −1.4 vs bf8; gradient energy 19× lower). So no bf4 tier is included: bf8
is the floor, and the only real perf lever here is activation dtype (bf8), not weight bit-width.

## Perf & quality (distilled 1080p, girl-singing, SEED=10, BH 4x8 galaxy)

| | Stage 1 | Stage 2 | **Total** |
|---|---|---|---|
| **bf16 (default)** | 2.47 | 2.92 | **7.05 s** |
| bf8 (`LTX_QUANT=all_bf8_lofi`) | 2.19 | 2.51 | **6.35 s** |

Opting into bf8 is ~0.70 s (~10%) faster. VBench-clean — with bf8 enabled all 1088-height floors clear,
on par with the bf16 default. The gate averages `VBENCH_SEEDS` clips (default 5); `dynamic_degree` is
near-binary per clip, so a single clip is too noisy to gate on:

| dim | bf8 | floor | bf16 (default) |
|---|---|---|---|
| subject_consistency | 0.979 | 0.92 | 0.985 |
| background_consistency | 0.974 | 0.93 | 0.961 |
| motion_smoothness | 0.991 | 0.955 | 0.994 |
| dynamic_degree | 1.00 | 0.8 | 0.80 |
| imaging_quality | 0.697 | 0.645 | 0.679 |

## Scope

The bf8 option is available for every LTX pipeline — `LTXDistilledPipeline`, `LTXOneStagePipeline`,
`LTXTwoStagesPipeline` all inherit `LTXPipeline.__init__`. All default to bf16; set
`LTX_QUANT=all_bf8_lofi` to enable bf8.

## Design & caching

The default (bf16) reads/writes the untagged transformer cache, byte-identical to `main`. Under
`LTX_QUANT=all_bf8_lofi` the resolved `QuantConfig` is baked into transformer **construction**: each DiT
linear is built at its preset weight dtype, with the activation cast, output pin, and compute-kernel /
ring-SDPA configs set in `__init__`. So `Parameter.load` typecasts weights direct-to-bf8 and the strict
tensorbin loader sees the dtype it expects — **no post-load hook**. A cache miss loads direct-to-quant
and writes quantized tensorbins; a cache hit (including `dynamic_load` reloads) reconstructs an
already-quantized module. `QuantConfig` lives beside the transformer model
(`models/transformers/ltx/quant_config.py`), so the model layer never imports up into `pipelines/`; the
pipeline resolves the preset before construction (`_resolve_quant_config`) and tags the bf8 cache
(`…q-all_bf8_lofi`) — a separate dir from the bf16 baseline, no cross-precision aliasing.

Device-validated (BH 4x8, `bh_4x8sp1tp0_ring`): under `LTX_QUANT=all_bf8_lofi` the bf8 cache materialized
and reloaded, rendering a correct girl-singing clip at 6.35 s; the bf16 default path is unchanged. The
construction-time refactor is byte-identical to the prior post-load-hook implementation (bf8 SEED=10
render md5-matched before/after).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:smarton/optimizer/ltx-bf8-denoise-port-2026-07-17)

### CODE REVIEW

File: models/tt_dit/pipelines/ltx/quant_config.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+    def _quant_self_attn(attn):
+        if attn is None:
+            return
+        _apply_linear_quant(attn.to_qkv, config.self_attn_qkv)
+        _apply_linear_quant(attn.to_out, config.self_attn_out)
+        attn.mm_compute_kernel_config = qkv_compute
+        attn.sdpa_compute_kernel_config = sdpa_compute
+        if sdpa_input_dtype is not None:
+            attn._sdpa_input_dtype = sdpa_input_dtype
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/quant_config.py]: Fixed in d402c4a0748. `apply_quant_config` now assigns `_sdpa_input_dtype` unconditionally (including None), so re-applying a preset that requests no SDPA cast clears a prior preset's cast instead of leaking it.

File: models/tt_dit/models/transformers/ltx/rope_ltx.py [hunk 2]
```diff
@@ -242,12 +242,22 @@ def prepare_video_rope(
...
+    there so the anchor carries the exact RoPE phase of the free grid frame it conditions."""
     v_shape = VideoLatentShape(batch=1, channels=128, frames=latent_frames, height=latent_height, width=latent_width)
     v_coords = video_get_patch_grid_bounds(v_shape)
     v_positions = get_pixel_coords(v_coords, scale_factors=(8, 32, 32), causal_fix=True).float()
     v_positions[:, 0, ...] = v_positions[:, 0, ...] / fps
+    if anchor_frames:
+        hw = latent_height * latent_width
+        v_positions = torch.cat(
+            [v_positions] + [v_positions[:, :, f * hw : (f + 1) * hw, :] for f in anchor_frames], dim=2
+        )
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/models/transformers/ltx/rope_ltx.py]: Fixed in d402c4a0748. Both `prepare_video_rope` and `prepare_av_cross_pe` now raise `ValueError` on anchor indices outside `[0, latent_frames)` before slicing, instead of silently building a length-mismatched embedding.

File: models/tt_dit/pipelines/ltx/quant_config.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            ffn_ff2=lc,
+            ring_sdpa=SDPAQuantConfig(),
+        )
+
+    @staticmethod
+    def all_bf8_lofi() -> QuantConfig:
+        """Weights bfloat8_b, LoFi compute. SDPA stays bf16/HiFi2.
+
+        ``activation_dtype`` is bf8 here but is inert unless ``LTX_QUANT_ACTIVATIONS`` is set, so by
+        default this preset is matmul-internal only: the collectives still move bf16.
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/quant_config.py]: Fixed in d402c4a0748. Corrected the docstring: `LTX_QUANT_ACTIVATIONS` defaults on, so activations are cast and the collectives move bf8 by default; set `LTX_QUANT_ACTIVATIONS=0` for a matmul-internal-only cast.

File: models/tt_dit/pipelines/ltx/pipeline_ltx.py [hunk 4]
```diff
@@ -608,6 +637,11 @@ def _prepare_transformer(self, idx: int = 0) -> None:
             is_fsdp=self.is_fsdp,
             lora_specs=state.lora_specs,
         )
+        # A registered quant preset typecasts the DiT-linear weights in place after each (re)load;
+        # a dynamic_load reload lands bf16, so this must run every call, not just the first.
+        hook = getattr(self, "_transformer_post_load_hook", None)
+        if hook is not None:
+            hook(state.model)
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/pipeline_ltx.py]: Fixed in b48847b3ba4. The transformer cache is now preset-tagged (`.q-all_bf8_lofi`) and holds bf8 tensorbins, and the quant config is applied to the built-but-unloaded module so `param.dtype` is bf8 *before* the strict load — a dynamic_load reload now reads bf8 into a bf8 param, no mismatch. Device-validated on BH 4x8: bf8 cache written (23G, marker-complete) and reloaded (one_stage bh_4x8sp1tp0_ring passed loading it).

File: models/tt_dit/pipelines/ltx/quant_config.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+# ---------------------------------------------------------------------------
+# Public API
+# ---------------------------------------------------------------------------
+
+
+def apply_quant_config_to_block(block, config: QuantConfig, arch, has_audio: bool) -> None:
+    """Quantize a single ``LTXTransformerBlock`` in place (weights + compute configs).
+
+    Factored out so the transformer-block PCC test can apply the exact same quant path the
+    pipeline uses against the diffusers torch oracle. ``arch`` is the device arch enum.
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/quant_config.py]: The shipped quant path is exercised end-to-end: the one_stage/two_stages/distilled pipeline tests default to `all_bf8_lofi` now, and the distilled test is VBench-gated, so a quant regression fails them. A dedicated block-level PCC unit test (apply `all_bf8_lofi`, assert the mapped weight dtypes + PCC vs the bf16 oracle) is a good faster-feedback follow-up — tracking it separately rather than expanding this PR.

File: models/tt_dit/models/transformers/ltx/transformer_ltx.py [hunk 6]
```diff
@@ -27,6 +27,53 @@
 from ....utils.tracing import traced_function
 from .attention_ltx import LTXAttention
 
+# gen#0 self-warms the DiT via a prep_run=True capture instead of the warmup denoise. Off by default:
+# it is only safe when the inner_step @traced_function captures with prep_run under this flag, which it
+# does not, so enabling it skips warmup with no self-warm and compiles the DiT cold in the reserved window.
+LTX_DIT_PREP_RUN = os.environ.get("LTX_DIT_PREP_RUN", "0") in ("1", "true", "True")
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/models/transformers/ltx/transformer_ltx.py]: Fixed in d402c4a0748. Removed `LTX_DIT_PREP_RUN` — it was never read (the prep_run capture path it described isn't wired), so it was a no-op switch.

File: models/tt_dit/pipelines/ltx/pipeline_ltx.py [hunk 7]
```diff
@@ -45,6 +45,12 @@
 
 LTX_UPSAMPLER_HF_REF = "Lightricks/LTX-2.3:ltx-2.3-spatial-upscaler-x2-1.1.safetensors"
 
+# Default DiT-linear quant preset. bf8 weights are the shipped 1080p tier: the perf targets and the
+# VBench floors are calibrated against it. LTX_QUANT="" opts back to the bf16 baseline; LTX_QUANT
+# names any other QuantConfig preset. The weight-cache name derives from the same value, so the
+# reader here and set_quant_config must resolve it identically or a run hits a stale-precision cache.
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/pipeline_ltx.py]: Fixed in b48847b3ba4. `cache_name` now appends `.q-<preset>`, so `LTX_QUANT` participates in the transformer cache key and the bf8 cache is a separate dir from the bf16 baseline. Updated the pipeline_ltx.py and quant_config.py doc comments to match.

**Review by cglagovich — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]

**Comment by smartonTT** [comment]
@friedrich please take a look at models/tt_dit/utils/cache.py.

It adds an optional post_load_hook to the cache load/save path. The hook (a bf8 quant typecast) runs on the module before the tensorbins are written and after every cache hit, so the cache stores post-hook weight dtypes and a reload stays consistent with the strict dtype check. Default post_load_hook=None is a no-op — no change for existing callers.

Thanks!

**Comment by friedrich** [comment]
@smartonTT I don't think `post_load_hook` is necessary. It seems to me like the model is already bf8 when the cache is written. And I would not call it after every cache load - cache loading already enforces the cache to have the correct dtype.

**Comment by smartonTT** [comment]
@friedrich how do you feel about it now? 
Btw, the previous solution was sort of how wan 2.2 did it... Maybe we want to clean wan up as well.

**Comment by friedrich** [comment]
Hi @smartonTT, I would prefer patching the weights before feeding them to the model, like you used to do. That way the change is much less invasive and does not increase complexity in the model code itself. Was there a reason to switch away from doing that? To be clear: the right point to do that would be before feeding the Torch state_dict to the model/cache.

File: models/tt_dit/pipelines/ltx/quant_config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/pipelines/ltx/quant_config.py]: Added `test_ltx_transformer_block_bf8` in `test_transformer_ltx.py` — the block-level bf8 test flagged as a follow-up above.

- **Mapped weight/activation dtypes:** builds `LTXTransformerBlock` under `all_bf8_lofi()` on the 4×8 ring and asserts the *loaded device-tensor* dtypes — projections (qkv/q/kv, ff1/ff2) bf8, the `to_out` fused-addcmul carve-out kept bf16, and the bf8 activation casts. A mis-wired role or a silent no-quant fails these asserts; note a missing activation cast can't be caught by PCC alone, since bf16 activations only *raise* it.
- **Block PCC:** vs the fp32 diffusers oracle — 99.93% on BH 4×8 (floor 0.99, a ring-noise margin below, matching the bf16 block test's convention).
- **Evict/reload:** structurally safe in the current design — no post-load hook; weights are bf8 at construction and the transformer cache is preset-tagged (`.q-all_bf8_lofi`), so a reload reads bf8 tensorbins into a bf8 param and the strict dtype check can't mismatch. Verified on device: the distilled AV run writes the bf8 cache and reloads it into bf8 params. Kept the new unit test focused on the mapping + numerics rather than duplicating that round-trip.

**Comment by smartonTT** [comment]
> Hi @smartonTT, I would prefer patching the weights before feeding them to the model, like you used to do. That way the change is much less invasive and does not increase complexity in the model code itself. Was there a reason to switch away from doing that? To be clear: the right point to do that would be before feeding the Torch state_dict to the model/cache.

Agents don't always pick the cleanest solution :D I think it's cleaner now. Let me know if you're good with it.

Agent's response: "There's no Torch bf8 — from_torch(dtype=…) is the quantizer and runs at Parameter construction, so a state_dict pre-pass can't produce bf8 weights. But your goal is met: all policy now lives in one LtxQuantProfile; the model modules just spread profile.linear_kwargs(role) and carry no dtype/fidelity literals."

**Comment by smartonTT** [comment]
@friedrich I'm merging since you're on vacation and I think I addressed your concerns, but feel free to request changes once you're back.
